doc/source/administrator/authentication.md

```diff
@@ -0,0 +1,427 @@
+      client_secret: your-client-secret
+      oauth_callback_url: https://your-jupyterhub-domain/hub/oauth_callback
+      authorize_url: https://${host}/auth/realms/${realm}/protocol/openid-connect/auth
+      token_url: https://${host}/auth/realms/${realm}/protocol/openid-connect/token
+      userdata_url: https://${host}/auth/realms/${realm}/protocol/openid-connect/userinfo
+    GenericOAuthenticator:
+      login_service: keycloak
+      username_key: preferred_username
+      userdata_params:
+        state: state
+    JupyterHub:
+      authenticator_class: oauthenticator.generic.GenericOAuthenticator
+```
+
+#### Auth0
+
+Auth0 is a popular commercial provider of identity management. The
+JupyterHub helm chart does not include support for Auth0 by default. To
+use Auth0, `extraEnv` and `extraConfig` must be configured as follows:
+
+Note that without the scope defined, authenticating to JupyterHub after
+already being logged in to Auth0 will fail.
+
+```yaml
+hub:
+  config:
+    OAuthenticator:
+      client_id: client-id-from-auth0-here
+      client_secret: client-secret-from-auth0-here
+      oauth_callback_url: https://your-jupyterhub-domain/hub/oauth_callback
+      scope:
+        - openid
+        - email
+    Auth0OAuthenticator:
+      auth0_subdomain: prod-8ua-1yy9
+    Authenticator:
+      admin_users:
+        - devops@example.com
+      auto_login: true
+    JupyterHub:
+      authenticator_class: oauthenticator.auth0.Auth0OAuthenticator
+```
+
+#### Full Example of Google OAuth2
+
+If your institution is a [G Suite customer](https://gsuite.google.com)
+that integrates with Google services such as Gmail, Calendar, and Drive,
+you can authenticate users to your JupyterHub using Google for
+authentication.
+
+1.  Log in to the [Google API
+    Console](https://console.developers.google.com).
+2.  Select a project > Create a project\... and set 'Project name'.
+    This is a short term that is only displayed in the console. If you
+    have already created a project you may skip this step.
+3.  Type "Credentials" in the search field at the top and click to
+    access the Credentials API.
+4.  Click "Create credentials", then "OAuth client ID". Choose
+    "Application type" > "Web application".
+5.  Enter a name for your JupyterHub instance. You can give it a
+    descriptive name or set it to be the hub's hostname.
+6.  Set "Authorized JavaScript origins" to be your hub's URL.
+7.  Set "Authorized redirect URIs" to be your hub's URL followed by
+    `/hub/oauth_callback`. For example,
+    `https://your-jupyterhub-domain/hub/oauth_callback`.
+8.  When you click "Create", the console will generate and display a
+    Client ID and Client Secret. Save these values.
+9.  Type "consent screen" in the search field at the top and click to
+    access the OAuth consent screen. Here you will customize what your
+    users see when they login to your JupyterHub instance for the first
+    time. Click Save when you are done.
```

Comment:
As an aside, the above steps would be really useful for non Z2JH users too! How do you feel about linking to this doc from other project docs?

Reply:
Sure why not. One answer would be to avoid locking in whats written exactly here by letting it be referenced by other projects.

I think i'd like to see some refactoring of this page in general, and preferably before referencing it since the structure may change a bit based on the refactoring.

As an example of refactoring, I think it could be nice to have a common section of the OAuthenticator configs, or link out to oauthenticator where we improve the docs of that etc. Not sure!

Reply:
I did some refactoring. I think in general, it would be good if we let OAuthenticator maintain the docs on how to setup client id etc, as it is used by z2jh / tljh and vanilla JupyterHub's etc, while z2jh is just one distribution where its relevant.